### ISSUE TITLE
TPM2 code does not return proper IV value to application/TSS stack

### ISSUE THREAD

**Issue by stefanberger** [stefanberger/libtpms#207]
I am filing this on behalf of  William Roberts <william.c.roberts@intel.com>

**Describe the bug**
libtpms added an OpenSSL back-end to the encryption code, however, the encryption routines are supposed to update the ivInOut paramater on successful execution of the AES cipher. However, the openssl backend code fails to do this, but when configured with --disable-use-openssl-functions, the crypto code goes back to the reference code implementation which properly updates the ivInOut. The side effect of not updating the ivInOut is that the output IV returned to applications is the same as the input IV. This results in weakened AES rounds.
Steps To Reproduce:
I have attached a script that uses the tpm2-software stack to reproduce the bug, as well as attached a fix patch.

1. Create an AES key with the TPM
2. Call EncryptDecrypt or EncryptDecrypt2 using that key.
3. Note the sent IV is same as returned IV.

*** Example of Running the Test Script **
```
./pcap-enc-dec.sh.sh swtpm
<snip>
IV in
00000000: dead beef dead beef dead beef dead beef  ................
IV out
00000000: dead beef dead beef dead beef dead beef  ................
Encrypted Result
00000000: 7642 3cc5 7eb8 4acb 6f9e 5fc8 f068 fdb4  vB<.~.J.o._..h..
```
Notice that IV in is equal to IV out.
After you apply the patch to libtpms, and rebuild swtpm, you can verify different output.

```
./pcap-enc-dec.sh.sh swtpm
<snip>
IV in
00000000: dead beef dead beef dead beef dead beef  ................
IV out
00000000: 211e 86fe 7199 8e7b f20f 4b95 6624 6bfc  !...q..{..K.f$k.
Encrypted Result
00000000: 211e 86fe 7199 8e7b f20f 4b95 6624 6bfc  !...q..{..K.f$k.
```

Note that the script can take two arguments:
First argument is the TCTI, like swtpm, device, or mssim.
Second argument, if specified must specify argument 1, AES mode of operation.
Also note, that I attached tpm2_log.pcap for a single run of that tool, you can open it in wireshark v3.0 or greater and see the TPM transmission bytes between the script and swtpm/libtpms.
Supporting Material/References: [Images are preferred. Videos (mp4) are allowed.]
See attached.
tpm2_log.pcap: pcap log file, can be opened with wireshark v3.0+.
pcap-enc-dec.sh.sh: reproduction script. Uses tpm2-software stack: https://github.com/tpm2-software
0001-CryptSym-fix-AES-output-IV.patch: fix patch to apply to libtpms
Access Vector Required for Exploitation:
Application data encrypted with multiple rounds of AES will be weakened, access to that cipher text could be compromised.
Vulnerability Exists in Default Configuration?:
Yes, you need to configure with --disable-use-openssl-functions to avoid it.
Is the exploitation trivial or does it involve a multi-step process that may depend on user/victim interaction?:
Depends.
Exploitation Requires Authentication?:
Depends on how the ciphertext is protected.
Under what privileges does the vulnerable service or component run?:
Access to ciphertext could be compromised.
Impact
Compromising encrypted data.

**Comment by stefanberger** [comment]
This issue has in the meantime been fixed with PRs #204, #203 , #202  applying the fix to stable-0.7.0, stable-0.8 and master.

**Comment by stefanberger** [comment]
This issue has been assigned CVE-2021-3446.
